### PR TITLE
Document symbols provider

### PR DESCRIPTION
**Pull request by pepeiborra**
Low hanging fruit, worth doing as VS Code makes document outlines actually useful by adorning them with warnings and errors markers.

### CODE REVIEW

File: src/Development/IDE/LSP/Outline.hs [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+      pure $ case mb_decls of
+        Nothing -> DSDocumentSymbols (List [])
+        Just (ParsedModule { pm_parsed_source = L _ltop HsModule { hsmodName, hsmodDecls, hsmodImports } })
+          -> let
+               declSymbols  = mapMaybe documentSymbolForDecl hsmodDecls
+               moduleSymbol = hsmodName <&> \(L l m) ->
+                 (defDocumentSymbol l :: DocumentSymbol)
+                   { _name  = pprText m
+                   , _kind  = SkFile
+                   , _range = Range (Position 0 0) (Position 10000000 0) -- _ltop is 0 0 0 0
```
Comment (aherrmann-da) [inline src/Development/IDE/LSP/Outline.hs]:
```suggestion
                   , _range = Range (Position 0 0) (Position maxBound 0) -- _ltop is 0 0 0 0
```
I'm guessing 10000000 is just meant to be larger than any actual file. Would `maxBound` be more meaningful?

File: src/Development/IDE/LSP/Outline.hs [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    , _kind = SkInterface
+    }
+documentSymbolForDecl (L l (DerivD DerivDecl { deriv_type })) =
+  gfindtype deriv_type <&> \(L (_ :: SrcSpan) name) ->
+    (defDocumentSymbol l :: DocumentSymbol) { _name = pprText @(HsType GhcPs)
+                                              name
+                                            , _kind = SkInterface
+                                            }
+documentSymbolForDecl (L l (ValD x)) =
+  gfindtype x <&> \(L (_ :: SrcSpan) name) ->
```
Comment (aherrmann-da) [inline src/Development/IDE/LSP/Outline.hs]: Bindings like `foo :: Int = 42` or `Just foo = Just 21` seem to be skipped. Maybe that's due to the use of `gfindtype`?

File: src/Development/IDE/LSP/Outline.hs [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    (defDocumentSymbol l :: DocumentSymbol)
+      { _name   = showRdrName name
+      , _kind   = case x of
+                    PatBind{} -> SkConstant
+                    _         -> SkFunction
```
Comment (aherrmann-da) [inline src/Development/IDE/LSP/Outline.hs]: I'm not sure the distinction between `SkConstant` and `SkFunction` makes sense at this point. As I understand, a binding like `Just plus2 = Just (+ 2)` would be qualified as `SkConstant`, while `two = 2` would be qualified as `SkFunction`. Sadly, LSP doesn't seem to have `SkValue`, but it seems more to correct to consistently use either `SkFunction` or `SkConstant`.

File: src/Development/IDE/LSP/Outline.hs (hunk 1 above)
Comment (pepeiborra) [reply src/Development/IDE/LSP/Outline.hs]: Good point, thanks!

File: src/Development/IDE/LSP/Outline.hs (hunk 2 above)
Comment (pepeiborra) [reply src/Development/IDE/LSP/Outline.hs]: Should be fixed now

**Pull request merged**